### PR TITLE
server,statistics: add dump_stats api for debug

### PR DESCRIPTION
**Pull request by fipped**
Add the HTTP API for dumping statistics, we can get the JSON type stats info by `dump_stats/{db}/{table}`

@lamxTyler @hanfei1991 @shenli  PTAL

to #5360

### CODE REVIEW

File: server/statistics_handler.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+package server
+
+import (
+	"github.com/gorilla/mux"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/model"
+	log "github.com/sirupsen/logrus"
+	"net/http"
```
Comment (lamxTyler) [inline server/statistics_handler.go]: Put the standard package to the upper.

File: server/statistics_handler.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+// StatsHandler is the handler for dump statistics
```
Comment (lamxTyler) [inline server/statistics_handler.go]: Add dots at the end of the line.

File: server/statistics_handler.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+type StatsHandler struct {
+	store kv.Storage
+	do    *domain.Domain
+}
+
+func (s *Server) newStatsHandler() *StatsHandler {
+	store, ok := s.driver.(*TiDBDriver)
+	if !ok {
+		panic("Invalid KvStore with illegal driver")
```
Comment (lamxTyler) [inline server/statistics_handler.go]: Wrong comment?

File: server/statistics_handler.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	var tikvStore kv.Storage
+	if tikvStore, ok = store.store.(kv.Storage); !ok {
```
Comment (lamxTyler) [inline server/statistics_handler.go]: No need to convert it, since `store.store` is `kv.Storage`.

File: statistics/table.go [hunk 5]
```diff
@@ -53,6 +54,28 @@ type Table struct {
 	Pseudo      bool
 }
 
+// JSONTable is used for Dumping statistics
```
Comment (lamxTyler) [inline statistics/table.go]: Add dot.

File: statistics/table.go [hunk 6]
```diff
@@ -53,6 +54,28 @@ type Table struct {
...
+type JSONTable struct {
+	DatabaseName string                 `json:"database_name"`
+	TableName    string                 `json:"table_name"`
+	Columns      map[string]*jsonColumn `json:"columns"` // Columns[col_name]
```
Comment (lamxTyler) [inline statistics/table.go]: The comment is not necessary?

File: statistics/table.go [hunk 7]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
 	}
 	return rowCount
 }
+
+// ConvertTo converts Bucket'LowerBound and UpperBound to type t
```
Comment (lamxTyler) [inline statistics/table.go]: `Bucket's` ?

File: statistics/table.go [hunk 8]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+func (b *Bucket) ConvertTo(h *Handle, t *types.FieldType) (err error) {
+	b.LowerBound, err = b.LowerBound.ConvertTo(h.ctx.GetSessionVars().StmtCtx, t)
+	if err != nil {
+		return
```
Comment (lamxTyler) [inline statistics/table.go]: Should be `error.Trace(err)`.

File: statistics/table.go [hunk 9]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+	}
+	b.UpperBound, err = b.UpperBound.ConvertTo(h.ctx.GetSessionVars().StmtCtx, t)
+	return err
+}
+
+func (b *Bucket) calScalar() {
```
Comment (lamxTyler) [inline statistics/table.go]: It is too simple to be a function.

File: statistics/table.go [hunk 10]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+				for i := range idx.Histogram.Buckets {
+					idx.Histogram.Buckets[i].calScalar()
+				}
+				idx.Histogram.ID = idxInfo.ID
+				tbl.Indices[idxInfo.ID] = idx
+			}
+		}
+	}
+	return tbl, nil
+}
```
Comment (lamxTyler) [inline statistics/table.go]: Better put these in a separate file.

File: statistics/table.go [hunk 11]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+		Count:        tbl.Count,
+		ModifyCount:  tbl.ModifyCount,
+		Version:      tbl.Version,
+	}
+	for key, val := range tbl.Columns {
+		hist := &Histogram{
+			ID:                val.Histogram.ID,
+			NDV:               val.Histogram.NDV,
+			NullCount:         val.Histogram.NullCount,
+			LastUpdateVersion: val.Histogram.LastUpdateVersion,
```
Comment (lamxTyler) [inline statistics/table.go]: `NullCount` and `LastUpdateVersion` is missing in the json?

File: statistics/table.go [hunk 12]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+		Count:       jsonTbl.Count,
+		Version:     jsonTbl.Version,
+		ModifyCount: jsonTbl.ModifyCount,
+	}
+	for key, val := range jsonTbl.Columns {
+		if loadAll || val.AlreadyLoad {
+			for _, colInfo := range tableInfo.Columns {
+				if colInfo.Name.L == key {
+					col := &Column{
+						Histogram: *HistogramFromProto(val.Histogram),
```
Comment (lamxTyler) [inline statistics/table.go]: Code blocks nested too deep...

File: statistics/table_test.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+	tk.MustExec("analyze table t")
+
+	is := s.do.InfoSchema()
+	h := s.do.StatsHandle()
+
+	tableInfo, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	jsonTbl, err := h.DumpStatsToJSON("test", tableInfo.Meta())
+	c.Assert(err, IsNil)
+	loadTbl, err := h.LoadStatsFromJSON(tableInfo.Meta(), jsonTbl, true)
```
Comment (lamxTyler) [inline statistics/table_test.go]: Do we need to check when the `loadAll` is false?

File: statistics/table.go [hunk 14]
```diff
@@ -438,3 +461,128 @@ func getPseudoRowCountByIntRanges(intRanges []ranger.IntColumnRange, tableRowCou
...
+	jsonTbl := &JSONTable{
+		DatabaseName: dbName,
+		TableName:    tableInfo.Name.L,
+		Columns:      make(map[string]*jsonColumn, len(tbl.Columns)),
+		Indices:      make(map[string]*jsonIndex, len(tbl.Indices)),
+		Count:        tbl.Count,
+		ModifyCount:  tbl.ModifyCount,
+		Version:      tbl.Version,
+	}
+	for key, val := range tbl.Columns {
```
Comment (lamxTyler) [inline statistics/table.go]: Better use more concrete name, like `id, col`.

File: server/statistics_handler.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/model"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+)
+
+// StatsHandler is the handler for dump statistics
+type StatsHandler struct {
+	store kv.Storage
```
Comment (shenli) [inline server/statistics_handler.go]: What's this `store` used for?

**Comment by shenli** [comment]
@fipped Please address the comments.

**Comment by fipped** [comment]
New PR #5534

**Pull request closed without merging**